### ISSUE TITLE
better use of hugging face infrastructure

### ISSUE THREAD

**Issue by jeremymanning** [ContextLab/llmXive#1241]
for storage and compute: we need to make better use of the huggingface infrastructure. e.g., hf-mount (https://github.com/huggingface/hf-mount) could be a great way of handling very large datasets in resource constrained environments.

stream (https://huggingface.co/docs/datasets/en/stream) is also a nice way of handling large datasets.

we could also use zerogpu (https://huggingface.co/docs/hub/en/spaces-zerogpu) for GPU compute

it could be useful to create (and test) tools that agents can then use to more effectively research

**Comment by jeremymanning** [comment]
I did a code-level spike on this. The main finding is that we should not treat “use Hugging Face infrastructure” as one integration: there are three distinct layers with different operational and reproducibility properties.

1. **Bounded dataset discovery/inspection** should be an llmXive library + MCP tool.
2. **Lazy filesystem access (`hf-mount`)** should be an opt-in execution adapter for agents and generated experiments.
3. **ZeroGPU** should be a narrowly scoped remote-compute backend, not a transparent replacement for arbitrary pipeline execution.

## What is already here

There is more groundwork than the issue description suggests:

- `pyproject.toml` already includes `huggingface_hub`.
- `librarian/dataset_sources.py::search_huggingface()` uses `HfApi`, chooses a real data file, and returns a stable Hub `resolve/main/...` URL.
- `librarian/dataset_resolver.py` downloads at most 256 KiB for format verification and intentionally preserves the stable Hub URL rather than an expiring signed CAS redirect.
- Planner/tasker/implementer prompts already recommend `load_dataset(..., streaming=True)` or `hf_hub_download`.
- `mcp_server/server.py` is already designed as thin wrappers around llmXive functions, so it is the right place to expose Hub capabilities to agents.

The missing piece is an executable, bounded abstraction. Today agents receive advice and URLs but cannot ask llmXive “show me the splits/schema/first five records” or “give me a lazy path to this repository.”

## Recommended architecture

### A. Start with Dataset Viewer tools (lowest-risk, highest immediate value)

Add a small `llmxive.huggingface` module and expose:

```python
inspect_hf_dataset(
    dataset: str,
    config: str | None = None,
    split: str | None = None,
    offset: int = 0,
    rows: int = 5,
) -> dict
```

Use the official Dataset Viewer `/splits` and `/rows` APIs rather than adding `datasets` to the core install. The API supports slices of at most 100 rows; llmXive should impose a tighter MCP cap such as 20 and truncate large cells. The viewer also offers `/size`, `/statistics`, and `/parquet`, which can be added after the first tool proves useful.

This complements rather than replaces the current resolver:

- resolver: discover and verify candidate sources;
- inspector: let an agent validate schema/content/size before planning;
- generated research code: use `datasets` streaming or Parquet directly for the actual analysis.

Hugging Face documents that Dataset Viewer can list configs/splits, return bounded row slices, report sizes, and expose converted Parquet shards:
- https://huggingface.co/docs/dataset-viewer/en/quick_start
- https://huggingface.co/docs/dataset-viewer/en/splits
- https://huggingface.co/docs/dataset-viewer/rows
- https://huggingface.co/docs/dataset-viewer/size
- https://huggingface.co/docs/dataset-viewer/parquet

I verified this minimal implementation against a real public dataset:

```python
import os
import requests

BASE = "https://datasets-server.huggingface.co"

def _get(endpoint, params):
    headers = {"User-Agent": "llmxive-hf-tools/0.1"}
    if token := os.environ.get("HF_TOKEN"):
        headers["Authorization"] = f"Bearer {token}"
    r = requests.get(
        f"{BASE}/{endpoint}", params=params, headers=headers, timeout=20
    )
    r.raise_for_status()
    return r.json()

def preview_hf_dataset(dataset, *, config=None, split=None, offset=0, rows=5):
    if not dataset or "/" not in dataset:
        raise ValueError("dataset must be a Hub id like 'owner/name'")
    if not 1 <= rows <= 20 or offset < 0:
        raise ValueError("rows must be 1..20 and offset must be non-negative")

    choices = _get("splits", {"dataset": dataset})["splits"]
    matching = [
        x for x in choices
        if (config is None or x["config"] == config)
        and (split is None or x["split"] == split)
    ]
    if not matching:
        available = [
            {"config": x["config"], "split": x["split"]} for x in choices
        ]
        raise ValueError(f"config/split not found; available={available}")

    selected = matching[0]
    page = _get("rows", {
        "dataset": dataset,
        "config": selected["config"],
        "split": selected["split"],
        "offset": offset,
        "length": rows,
    })
    return {
        "dataset": dataset,
        "config": selected["config"],
        "split": selected["split"],
        "num_rows_total": page.get("num_rows_total"),
        "features": [f["name"] for f in page.get("features", [])],
        "rows": [r["row"] for r in page.get("rows", [])],
        "partial": page.get("partial", False),
    }
```

Direct run:

```text
dataset: cornell-movie-review-data/rotten_tomatoes
selected: default/train
requested/returned rows: 2/2
num_rows_total: 8530
features: [text, label]
invalid config: rejected with available config/split choices
exit: 0
```

Production code should additionally recursively truncate strings/arrays, translate network exceptions into the MCP server’s `{"error": ...}` convention, and never echo tokens.

For actual large-scale iteration, Hugging Face documents that `load_dataset(..., streaming=True)` can begin using a 45 TB split immediately, supports column/filter pushdown for Parquet, and returns an `IterableDataset`; importantly, iterable datasets are unsuitable for random access. `take`, sharding, bounded-buffer shuffle, and checkpoint state are available:
https://huggingface.co/docs/datasets/stream

The Datasets paper describes the library’s Arrow-backed, interoperable architecture and provenance/versioning motivation:
https://aclanthology.org/2021.emnlp-demo.21/

### B. Add a lazy-filesystem abstraction before making `hf-mount` mandatory

There are two useful implementations:

1. **In-process default:** `huggingface_hub.HfFileSystem`, already available through the current dependency. It exposes `hf://datasets/...` paths and works with fsspec-aware libraries. HF warns that the fsspec compatibility layer has overhead and recommends direct `HfApi` methods where possible:
   https://huggingface.co/docs/huggingface_hub/package_reference/hf_file_system
2. **Executor opt-in:** `hf-mount`, when the executable and an NFS/FUSE-capable environment are present.

I verified lazy `HfFileSystem` access with the current llmXive environment:

```python
from huggingface_hub import HfFileSystem

fs = HfFileSystem(token=False)
root = "datasets/cornell-movie-review-data/rotten_tomatoes"
print(fs.ls(root, detail=False))
with fs.open(f"{root}/README.md", "rt") as handle:
    print(handle.read(80))
```

This listed five remote files and read only the requested README prefix, exit 0.

`hf-mount` is promising specifically for generated code and shell-oriented agents: it exposes repos/buckets through NFS or FUSE, fetches bytes lazily, supports read-only dataset repos, and requires no root with the recommended NFS backend. However, its own README says it is not suitable for latency-sensitive random I/O, strong-consistency workloads, multi-writer storage, or cross-client locking; files can be stale for up to 10 seconds:
https://github.com/huggingface/hf-mount

Therefore add a context-managed adapter rather than embedding raw mount commands in prompts:

```python
with mount_hf_repo(
    "datasets/open-index/hacker-news",
    revision="<commit SHA>",
) as path:
    run_research_code(data_root=path)
```

Implementation requirements:

- probe `shutil.which("hf-mount")`; return a clear capability error if absent;
- prefer NFS and a private directory under the project’s scratch area;
- pass credentials via environment, never command-line arguments or logs;
- require a revision/commit SHA in reproducible pipeline stages;
- start with read-only repos only;
- ensure unmount/daemon cleanup in `finally`, including cancellation;
- record repo ID, revision, backend, and mounted subfolder in experiment provenance;
- retain `HfFileSystem`/streaming fallback for containers where mounts are unavailable.

For complete downloads, continue to use `hf_hub_download`/`snapshot_download`: HF’s version-aware cache avoids repeated downloads and supports revision pinning:
https://huggingface.co/docs/huggingface_hub/guides/download
https://huggingface.co/docs/huggingface_hub/guides/manage-cache

For mutable checkpoints/logs/intermediate artifacts, evaluate Hub Storage Buckets rather than Dataset repos. Buckets are deliberately mutable and non-versioned, while repos retain history and cards:
https://huggingface.co/docs/hub/storage-buckets

A good policy is:

- Dataset repo + pinned revision: published/reproducible inputs and outputs.
- Bucket: scratch data, checkpoints, traces, transient artifacts.
- Local cache: repeated reads on a runner.
- `hf-mount`: lazy POSIX-like presentation where the executor supports it.

### C. Treat ZeroGPU as a queue-backed service boundary

ZeroGPU is currently Gradio-only and supports specific Python/PyTorch versions. GPU functions use `@spaces.GPU`, have a default 60-second allocation unless configured otherwise, and `xlarge` consumes twice the quota. It also does not support `torch.compile`:
https://huggingface.co/docs/hub/spaces-zerogpu

That makes it appropriate for bounded kernels such as:

- embeddings for a bounded batch;
- image/audio feature extraction;
- short inference/evaluation jobs;
- perhaps one deterministic analysis primitive with a typed schema.

It is not a good first target for arbitrary agent-generated Python, long training runs, or a transparent “remote CUDA” abstraction.

Every Gradio Space is callable as an API; clients can submit asynchronously, inspect queue status, cancel, and retrieve results. Authentication consumes the caller’s quota:
https://huggingface.co/docs/hub/spaces-api-endpoints

Proposed design:

```python
submit_hf_space_job(
    space_id: str,
    api_name: str,
    inputs: dict,
    timeout_s: int = 120,
    idempotency_key: str | None = None,
) -> {
    "status": "completed|queued|failed|timed_out",
    "result": ...,
    "space_revision": ...,
    "elapsed_s": ...,
}
```

Do not allow arbitrary `space_id` initially. Maintain an allowlist of llmXive-owned Spaces and endpoint schemas. Pin/record the Space commit, model revision, input hashes, and dependency lock. Use `Client.submit()` so a pipeline worker does not block blindly; propagate cancellation, impose wall-clock/output caps, and classify queue/quota exhaustion as transient rather than scientific failure.

## Staged implementation plan

### Phase 1: useful agent tools without new infrastructure

- Add `src/llmxive/huggingface/datasets.py`.
- Implement `list_hf_dataset_splits`, `inspect_hf_dataset`, and optionally `hf_dataset_size`.
- Reuse `requests`, `HF_TOKEN`, current error conventions, and 20-second timeouts.
- Add wrappers to `mcp_server/server.py`.
- Include the selected dataset ID/config/split and stable URL in planner context.
- Add an optional `hf` extra only when actual `datasets` iteration is needed:

```toml
hf = ["datasets>=<tested-minimum>", "gradio_client>=<tested-minimum>"]
```

Tests:

- mock `/splits` and `/rows`;
- auto-selection and explicit selection;
- invalid config/split enumerates choices;
- max rows, nonnegative offset, timeout/401/404/429 handling;
- nested string/list truncation and JSON serialization;
- `HF_TOKEN` is accepted but never returned/logged;
- MCP tool maps all failures to structured errors;
- opt-in live test behind `LLMXIVE_REAL_TESTS=1`.

### Phase 2: storage execution adapter

- Add `HfFileSystem` helpers for list/open/glob with bounded read sizes.
- Add an `hf-mount` capability probe and context manager behind a feature flag.
- Pin revisions; prefer read-only and NFS; guarantee cleanup.
- Add executor integration tests with a fake `hf-mount` binary/process, plus a manually gated live smoke test.
- Benchmark cold first read, warm read, bytes transferred, disk use, and cleanup on cancellation against `snapshot_download` and `datasets` streaming.

Acceptance criteria:

- a large repo can be inspected without a full clone;
- a bounded job reads only needed files/bytes;
- no orphan daemon/mount remains after success, exception, timeout, or SIGTERM;
- identical pinned revisions produce identical provenance;
- environments without mount support fall back cleanly.

### Phase 3: one owned ZeroGPU service

- Build one Gradio Space around a narrow, deterministic GPU primitive.
- Pin model and Space revisions and publish the input/output schema.
- Add an allowlisted async client with retry/backoff, timeout, cancellation, quota/queue telemetry, and output caps.
- Compare output against a CPU/local fixture within a documented tolerance.
- Only generalize after measuring queue latency, failure rate, reproducibility, and quota cost.

## Risks/open questions

- Do we have an HF organization plan that permits hosting ZeroGPU Spaces, and what quota/billing ceiling should pipeline runs enforce?
- Are any research datasets private, gated, export-controlled, or subject to data-residency/IRB rules? Tokens need least privilege and must never enter generated artifacts.
- Should mutable Hub buckets be enabled at all before write allowlists, lifecycle rules, and provenance are implemented?
- How should dataset licenses/cards be surfaced to the planner? Reachability and schema inspection do not establish permission or scientific suitability.
- Which environments can run NFS mounts? CI and restricted containers likely need `HfFileSystem`/streaming fallback.
- What maximum row/cell/output sizes keep MCP responses safe from context exhaustion and hostile dataset contents?
- Agent-visible dataset rows are untrusted content; prompts must mark them as data, not instructions.

## Suggested first PR

Keep the first PR deliberately small: `inspect_hf_dataset` + MCP wrapper + unit/live-gated tests + documentation. This immediately gives agents a useful verified tool, requires no daemon and no new core dependency, and creates the interface that later streaming, `hf-mount`, and ZeroGPU adapters can build on.
